include/xtensor/xio.hpp

```diff
@@ -87,280 +87,356 @@ namespace xt
 
-        auto signs_it = i_part.begin();
-        for (auto it = i_s.begin(); it != i_s.end(); it++)
+            private:
+                printer<value_type> real_printer, imag_printer;
+                cache_type m_signs;
+                cache_iterator m_it;
+        };
+
+        template <class T>
+        struct printer<T, std::enable_if_t<!std::is_fundamental<typename T::value_type>::value && !detail::is_complex<typename T::value_type>::value>>
         {
-            int j_idx = (*it).find(' ');
-            if (j_idx >= 0)
+            using value_type = typename T::value_type;
+            using cache_type = std::vector<std::string>;
+            using cache_iterator = typename cache_type::const_iterator;
+
+            void init(std::ostream& out)
             {
-                (*it).insert(std::size_t(j_idx), 1, 'j');
+                m_it = m_cache.cbegin();
             }
-            else
+
+            std::ostream& print_next(std::ostream& out)
             {
-                *it += 'j';
+                out << std::setw(m_width) << *m_it;
+                ++m_it;
+                return out;
             }
-            if (*signs_it < 0)
+
+            void update(const value_type& val)
             {
-                *it = '-' + *it;
+                std::stringstream buf;
+                buf << val;
+                std::string s = buf.str();
+                if(s.size() > m_width)
+                {
+                    m_width = s.size();
+                }
+                m_cache.push_back(s);
             }
-            else
+
+            int width()
             {
-                *it = '+' + *it;
+                return m_width;
             }
-            signs_it++;
-        }
-
-        xt::xarray<std::string> s = r_s + i_s;
 
-        std::cout << s << std::endl;
+            private:
+                int m_width = 0;                
+                cache_type m_cache;
+                cache_iterator m_it;
+        };
     }
 
     template <class E>
-    inline std::ostream& operator<<(std::ostream& out, const xexpression<E>& e)
+    std::ostream& pretty_print(const xexpression<E>& e, std::ostream& out = std::cout, unsigned int precision = 8)
     {
+        detail::printer<E> p;
+        const E& d = e.derived_cast();
+
         int lim = -1;
-        if (e.derived_cast().size() > 1000)
+        std::size_t sz = std::accumulate(d.shape().begin(), d.shape().end(), 1, std::multiplies<>());
+        if (sz > 1000)
         {
             int lim = 3;
         }
-        return detail::xout<5>::output(out, e.derived_cast(), 1, 3);
+
+        detail::recurser<5>::run(p, d, lim);
+        p.init(out);
+        return detail::xout<5>::output(out, d, p, 1, p.width(), lim);
+    }
+
+    template <class E>
+    inline std::ostream& operator<<(std::ostream& out, const xexpression<E>& e)
+    {
+        return pretty_print(e, out);
+    }
+
+    inline void set_precision(int prec, std::ostream& out = std::cout)
```

Comment:
I'm not sure we need a wrapper around ``out << std::setprecision(...)`` but maybe you have some use case ?